File: frontend/src/core/api/apiClient.ts
import axios, {
  AxiosError,
  type AxiosInstance,
  type AxiosRequestConfig,
  type InternalAxiosRequestConfig,
} from 'axios';

import { asRecord, parseApiResponse, type ApiResponse } from './contracts';
import { ApiError, AuthenticationRequiredError, AuthorizationError } from './errors';

export interface AuthSession {
  getAccessToken(): string | null;
  refreshAccessToken(): Promise<string | null>;
  authenticationRequired(): void;
}

type RetriableRequest = InternalAxiosRequestConfig & { q016Retried?: boolean };

export class ApiClient {
  private readonly axios: AxiosInstance;
  private refreshInFlight: Promise<string | null> | null = null;

  constructor(baseUrl: string, private readonly auth: AuthSession, instance?: AxiosInstance) {
    this.axios = instance ?? axios.create({ baseURL: baseUrl, timeout: 15_000 });
    this.axios.interceptors.request.use((request) => {
      const token = this.auth.getAccessToken();
      if (token && !request.headers.has('Authorization')) {
        request.headers.Authorization = `Bearer ${token}`;
      }
      return request;
    });
    this.axios.interceptors.response.use(
      (response) => response,
      async (error: unknown) => this.handleTransportError(error),
    );
  }

  async get<T>(path: string, parseData: (data: unknown) => T, config?: AxiosRequestConfig): Promise<T> {
    return this.unwrap(await this.axios.get<unknown>(path, config), parseData);
  }

  async post<T>(path: string, body: unknown, parseData: (data: unknown) => T): Promise<T> {
    return this.unwrap(await this.axios.post<unknown>(path, body), parseData);
  }

  private unwrap<T>(response: { data: unknown; status: number }, parseData: (data: unknown) => T): T {
    let envelope: ApiResponse<T>;
    try {
      envelope = parseApiResponse(response.data, parseData);
    } catch {
      throw new ApiError(
        'INVALID_API_RESPONSE',
        'The BrokerOS Risk API returned an invalid response.',
        response.status,
      );
    }
    if (envelope.code !== 'SUCCESS' || envelope.data === null) {
      throw new ApiError(envelope.code, envelope.message, response.status);
    }
    return envelope.data;
  }

  private async handleTransportError(error: unknown): Promise<never> {
    if (!(error instanceof AxiosError) || !error.config) {
      throw new ApiError('API_UNAVAILABLE', 'The BrokerOS Risk API is unavailable.', null);
    }

    const request = error.config as RetriableRequest;
    const status = error.response?.status ?? null;
    if (status === 401 && !request.q016Retried) {
      request.q016Retried = true;
      const refreshed = await this.refreshOnce();
      if (refreshed) {
        request.headers.Authorization = `Bearer ${refreshed}`;
        return this.axios.request(request);
      }
      this.auth.authenticationRequired();
      throw new AuthenticationRequiredError(this.failureMessage(error) ?? undefined);
    }

    const failure = this.failureEnvelope(error);
    if (status === 403) {
      throw new AuthorizationError(failure?.code, failure?.message);
    }
    throw new ApiError(
      failure?.code ?? 'API_UNAVAILABLE',
      failure?.message ?? 'The BrokerOS Risk API is unavailable.',
      status,
    );
  }

  private refreshOnce(): Promise<string | null> {
    if (!this.refreshInFlight) {
      this.refreshInFlight = this.auth.refreshAccessToken().finally(() => {
        this.refreshInFlight = null;
      });
    }
    return this.refreshInFlight;
  }

  private failureEnvelope(error: AxiosError): { code: string; message: string } | null {
    try {
      const record = asRecord(error.response?.data, 'ApiResponse');
      return {
        code: typeof record.code === 'string' ? record.code : 'API_ERROR',
        message:
          typeof record.message === 'string' && record.message.trim()
            ? record.message
            : 'The request could not be completed.',
      };
    } catch {
      return null;
    }
  }

  private failureMessage(error: AxiosError): string | null {
    return this.failureEnvelope(error)?.message ?? null;
  }
}

File: frontend/lib/core/api/api_client.dart
import 'package:dio/dio.dart';

import 'api_contract.dart';

abstract interface class AuthTokenDelegate {
  String? get accessToken;

  Future<String?> refresh();
}

class ApiClient {
  ApiClient({
    required String baseUrl,
    required AuthTokenDelegate auth,
    Dio? dio,
  }) : _dio = dio ?? Dio(BaseOptions(baseUrl: baseUrl)) {
    _dio.interceptors.add(_BearerInterceptor(_dio, auth));
  }

  final Dio _dio;

  Future<ApiResult<T>> get<T>(
    String path, {
    Map<String, dynamic>? queryParameters,
    required T Function(Object? json) decode,
  }) async {
    return _request(
      () => _dio.get<Object?>(path, queryParameters: queryParameters),
      decode,
    );
  }

  Future<ApiResult<T>> post<T>(
    String path, {
    Object? body,
    required T Function(Object? json) decode,
  }) async {
    return _request(() => _dio.post<Object?>(path, data: body), decode);
  }

  Future<ApiResult<T>> _request<T>(
    Future<Response<Object?>> Function() send,
    T Function(Object? json) decode,
  ) async {
    try {
      final response = await send();
      final body = _asMap(response.data);
      final envelope = ApiEnvelope<T>.fromJson(body, (json) {
        if (json == null) {
          return null;
        }
        return decode(json);
      });
      if (envelope.code == ResultCode.success && envelope.data != null) {
        return ApiSuccess<T>(envelope.data as T);
      }
      return ApiError<T>(ApiFailure(
        code: envelope.code,
        message: envelope.message,
        httpStatus: response.statusCode,
      ));
    } on DioException catch (error) {
      final response = error.response;
      final data = response?.data;
      if (data is Map) {
        final envelope = ApiEnvelope<Object?>.fromJson(
          data.cast<String, dynamic>(),
          (json) => json,
        );
        return ApiError<T>(ApiFailure(
          code: envelope.code,
          message: envelope.message,
          httpStatus: response?.statusCode,
        ));
      }
      return ApiError<T>(ApiFailure(
        code: response?.statusCode == 403
            ? ResultCode.authorizationDenied
            : ResultCode.unknown,
        message: response?.statusCode == 403
            ? 'You are not authorized to perform this operation.'
            : 'The BrokerOS Risk API is unavailable.',
        httpStatus: response?.statusCode,
      ));
    } on FormatException catch (error) {
      return ApiError<T>(ApiFailure(
        code: ResultCode.unknown,
        message: 'The API returned an invalid response: ${error.message}',
        httpStatus: null,
      ));
    }
  }

  Map<String, dynamic> _asMap(Object? value) {
    if (value is! Map) {
      throw const FormatException('expected an object envelope');
    }
    return value.cast<String, dynamic>();
  }
}

class _BearerInterceptor extends Interceptor {
  _BearerInterceptor(this._dio, this._auth);

  final Dio _dio;
  final AuthTokenDelegate _auth;

  @override
  void onRequest(RequestOptions options, RequestInterceptorHandler handler) {
    final token = _auth.accessToken;
    if (token != null) {
      options.headers['Authorization'] = 'Bearer $token';
    }
    handler.next(options);
  }

  @override
  Future<void> onError(
    DioException error,
    ErrorInterceptorHandler handler,
  ) async {
    final options = error.requestOptions;
    if (error.response?.statusCode == 401 && options.extra['q016Retried'] != true) {
      final refreshed = await _auth.refresh();
      if (refreshed != null) {
        options.extra['q016Retried'] = true;
        options.headers['Authorization'] = 'Bearer $refreshed';
        try {
          handler.resolve(await _dio.fetch<Object?>(options));
          return;
        } on DioException {
          // The original typed API failure is returned below without token data.
        }
      }
    }
    handler.next(error);
  }
}
